### PR TITLE
Feature/appscan enterprise remote

### PR DESCRIPTION
**Pull request by geoffwhittington**

### CODE REVIEW

**Comment by geoffwhittington** [comment]
@emintham do you have the bandwidth to code review this?

File: sdetools/alm_integration/alm_plugin_base.py [hunk 1]
```diff
@@ -863,3 +863,59 @@ def translate_priority(self, priority):
             else:
                 if int(key) == priority:
                     return pmap[key]
+
+
+class MacroAlmConnector(AlmConnector):
```
Comment (emintham) [inline sdetools/alm_integration/alm_plugin_base.py]: Missing a docstring.

File: sdetools/modules/import_appscan/appscan_integrator.py [hunk 2]
```diff
@@ -3,22 +3,55 @@
...
 
 
+class AppScanAPI(RESTBase):
+
+    def __init__(self, config):
+        self.session_ID = None
+        super(AppScanAPI, self).__init__('analysis', 'AppScan', config)
+
+    def start_session(self):
+        logger.info('Loging in as %s' % self.config['analysis_user'])
```
Comment (emintham) [inline sdetools/modules/import_appscan/appscan_integrator.py]: Logging

File: sdetools/modules/import_appscan/appscan_integrator.py [hunk 3]
```diff
@@ -3,22 +3,55 @@
...
+            "featureKey": "AppScanEnterpriseUser",
+            "clientVersion": "",
+            "clientIp": "",
+            "clientHostName": ""
+        }
+        try:
+            session = self.call_api(target_url, method=URLRequest.POST, args=credentials, auth_mode='cookie')
+        except APIError, err:
+            raise AppScanIntegrationError("Failed to reach server. %s" % err)
+        self.session_info = {'sessionId': session['sessionId']}
```
Comment (emintham) [inline sdetools/modules/import_appscan/appscan_integrator.py]:
```python
self.session_info = {
    'sessionId': session['sessionId'],
    'session-cookie-name': "asc_sso_token",
    'session-token': "asc_sso_token"
}
```

Is there a reason why sessionId has camel case while the others have hyphens?

File: sdetools/modules/import_appscan/appscan_integrator.py [hunk 4]
```diff
@@ -77,3 +133,71 @@ def _make_finding(self, item):
 
     def generate_findings(self):
         return [self._make_finding(item) for item in self.findings]
+
+    def translate(self, json):
+        items = []
+        for item in json:
+            items.append({'id': item[ self.issue_columns['issuetype'] ], 'description': item[ self.issue_columns['issuetype'] ], 'count': 1})
```
Comment (emintham) [inline sdetools/modules/import_appscan/appscan_integrator.py]: issuetype for both id and description?

```python
return [{
    'id': item[self.issue_columns['issuetype']],
    'description': item[self.issue_columns['issuetype']],
    'count': 1
} for item in json]

File: sdetools/modules/import_appscan/tests/appscan_test.py [hunk 5]
```diff
@@ -1,19 +1,44 @@
...
+from sdetools.modules.import_appscan.appscan_integrator import AppScanIntegrator, AppScanAPI
+from appscan_response_generator import AppScanResponseGenerator
+from sdetools.sdelib.commons import get_directory_of_current_module
+from sdetools.sdelib.testlib.mock_response import MOCK_ALM_RESPONSE, MOCK_SDE_RESPONSE
 
 
 class TestAppScanIntegration(BaseIntegrationTest, unittest.TestCase):
     @classmethod
-    def setUpClass(cls):
+    def setUpClass(cls):
```
Comment (emintham) [inline sdetools/modules/import_appscan/tests/appscan_test.py]: extra space at the end

File: sdetools/modules/sync_rally/rally_plugin.py [hunk 6]
```diff
@@ -299,20 +315,34 @@ def alm_validate_configurations(self):
...
+        if card_type_details['field_validation_state'] not in self.artifact_fields:
+             raise AlmException('Unable to locate the project status field: %s' % card_type_details['name'])
+
+        # Collect the valid issue states
+        state_attribute = self.artifact_fields[card_type_details['field_validation_state']]
+        allowed_values = [v['StringValue'] for v in state_attribute['AllowedValues']]
+
+        # Validate the new status
+        if self.config['rally_new_status'] not in allowed_values:
+            raise AlmException('Invalid rally_new_status: %s. Expected one of %s' %
```
Comment (emintham) [inline sdetools/modules/sync_rally/rally_plugin.py]: You can just do
```python
raise AlmException('Invalid rally_new_status: %s. Expected one of %s',
                   self.config['rally_new_status'],
                   ', '.join(allowed_values))
```

File: sdetools/modules/sync_rally/rally_plugin.py [hunk 7]
```diff
@@ -299,20 +315,34 @@ def alm_validate_configurations(self):
...
+            raise AlmException('Invalid rally_new_status: %s. Expected one of %s' % 
+                               (self.config['rally_new_status'], ', '.join(allowed_values)))
 
-                if self.config['rally_new_status'] not in allowed_values:
-                    raise AlmException('Invalid rally_new_status "%s". Expected one of %s' %
-                                       (self.config['rally_new_status'], allowed_values))
+        # Validate the done stasuses
+        for done_status in self.config['rally_done_statuses']:
+            if done_status not in allowed_values:
+                raise AlmException('Invalid rally_done_statuses value: %s. Expected one of %s' %
```
Comment (emintham) [inline sdetools/modules/sync_rally/rally_plugin.py]: same here.

File: sdetools/modules/sync_rally/rally_plugin.py [hunk 8]
```diff
@@ -299,20 +315,34 @@ def alm_validate_configurations(self):
...
-                if difference_set:
-                    raise AlmException('Invalid rally_done_statuses %s. Expected one of %s' %
-                                       (difference_set, allowed_values))
-                return
-        raise AlmException('Unable to retrieve allowed values for "%s"' % card_type_details['name'])
+        # Validate user-defined custom fields are in the project
+        unexpected_custom_fields = [x for x in self.config['alm_custom_fields'].keys()
+                                    if x not in self.artifact_fields.keys()]
+        if unexpected_custom_fields:
+            raise AlmException('Specified fields not found in the project: %s' % ', '.join(unexpected_custom_fields))
```
Comment (emintham) [inline sdetools/modules/sync_rally/rally_plugin.py]: As above

File: sdetools/modules/sync_rally/rally_plugin.py [hunk 9]
```diff
@@ -325,7 +355,7 @@ def rally_get_artifact(self, name, query, card_type, artifact_type, api):
         try:
             result = self.alm_plugin.call_api('%s.js' % api, args=query_args)
         except APIError, err:
-            raise AlmException('Unable to get %s %s from Rally. Reason: %s' % (card_type, name, str(err)))
+            raise AlmException('Unable to get %s %s. Reason: %s' % (card_type, name, str(err)))
```
Comment (emintham) [inline sdetools/modules/sync_rally/rally_plugin.py]: As above

File: sdetools/modules/sync_rally/rally_plugin.py [hunk 10]
```diff
@@ -335,7 +365,7 @@ def rally_get_artifact(self, name, query, card_type, artifact_type, api):
         try:
             task_data = self.alm_plugin.call_api(task_result_url)
         except APIError, err:
-            raise AlmException('Unable to get artifact from Rally. Reason: %s' % (str(err)))
+            raise AlmException('Unable to get artifact. Reason: %s' % (str(err)))
```
Comment (emintham) [inline sdetools/modules/sync_rally/rally_plugin.py]: As above

File: sdetools/modules/sync_rally/rally_plugin.py [hunk 11]
```diff
@@ -375,7 +405,7 @@ def _apply_labels_if_needed(self, labels, artifact_data):
         try:
             self.alm_plugin.call_api(artifact_id, args=tag_update, method=self.alm_plugin.URLRequest.POST)
         except APIError, err:
-            raise AlmException('Unable to update tag %s for artifact %s in Rally because of %s' %
+            raise AlmException('Unable to update tag %s for artifact %s because of %s' %
```
Comment (emintham) [inline sdetools/modules/sync_rally/rally_plugin.py]: As above

File: sdetools/modules/sync_rally/rally_plugin.py [hunk 12]
```diff
@@ -434,22 +471,16 @@ def alm_add_task(self, task):
...
-            for key in self.config['alm_custom_fields']:
-                create_args[card_type_details['type']][key] = self.config['alm_custom_fields'][key]
-
         try:
             result = self.alm_plugin.call_api('%s/create.js' % card_type_details['api'],
                                               method=self.alm_plugin.URLRequest.POST, args=create_args)
         except APIError, err:
-            raise AlmException('Unable to add task to Rally %s because of %s' %
-                               (task['id'], err))
+            raise AlmException('Unable to add task %s because of %s' % (task['id'], err))
```
Comment (emintham) [inline sdetools/modules/sync_rally/rally_plugin.py]: As above

File: sdetools/modules/sync_rally/rally_plugin.py [hunk 13]
```diff
@@ -434,22 +471,16 @@ def alm_add_task(self, task):
...
 
         logger.debug('Task %s added to Rally Project', task['id'])
 
         if result['CreateResult']['Errors']:
-            raise AlmException('Unable to add task to Rally %s. Reason: %s' %
+            raise AlmException('Unable to add task %s. Reason: %s' %
```
Comment (emintham) [inline sdetools/modules/sync_rally/rally_plugin.py]: As above

File: sdetools/modules/sync_rally/tests/rally_plugin_test.py [hunk 14]
```diff
@@ -41,16 +41,16 @@ def test_invalid_new_status(self):
         allowed_statuses = [u'Defined', u'In-Progress', u'Completed', u'Accepted']
         self.config['rally_new_status'] = bad_status
 
-        self.assert_exception(AlmException, '', 'Invalid rally_new_status "%s". Expected one of %s' %
-                              (bad_status, allowed_statuses), self.connector.alm_connect)
+        self.assert_exception(AlmException, '', 'Invalid rally_new_status: %s. Expected one of %s' %
```
Comment (emintham) [inline sdetools/modules/sync_rally/tests/rally_plugin_test.py]: As above

File: sdetools/modules/sync_rally/rally_plugin.py [hunk 15]
```diff
@@ -499,7 +530,7 @@ def alm_update_task_status(self, task, status):
         try:
             self.alm_plugin.call_api(task.get_alm_task_ref(), args=trans_args, method=self.alm_plugin.URLRequest.POST)
         except APIError, err:
-            raise AlmException('Unable to update task status to %s for artifact %s in Rally because of %s' %
+            raise AlmException('Unable to update task status to %s for artifact %s because of %s' %
```
Comment (emintham) [inline sdetools/modules/sync_rally/rally_plugin.py]: As above

**Comment by geoffwhittington** [comment]
Thanks for the review - I branched this off the old rally PR and forgot to rebase. Interesting what you can find the second time around :)

**Comment by geoffwhittington** [comment]
@emintham Github looks broken - the jira/rally changes have already been merged into master

File: sdetools/modules/import_appscan/appscan_integrator.py (hunk 3 above)
Comment (geoffwhittington) [inline sdetools/modules/import_appscan/appscan_integrator.py]: I removed the last two keys and the integration still works - problem solved :)

File: sdetools/modules/import_appscan/appscan_integrator.py (hunk 4 above)
Comment (geoffwhittington) [inline sdetools/modules/import_appscan/appscan_integrator.py]: Yeah we need to identify the weakness and have something to show the user - no guarantee there's anything else than the issue type

**Pull request merged**